### DISCUSSION TITLE
Using uasyncio server as HTTP Server: closing trouble

### DISCUSSION THREAD

**Discussion by milo1987** [micropython/micropython#10745] [Using MicroPython] [answered]
Hello Micropython community.
I hope you can help me.

I'm new on Python and programming for a esp8266 and esp32 device. I'm programming a smartdevices, that connects to iobroker and has a webinterface for config.
I'm using a few loops from uasyncio, that handle the different tasks.

One of them is a uasyncio.start_server instance, which calls my function httpLoop:

````
async def httpLoop(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        try:
            self.ms.log("HTTP-Client Connected...")
            s = str(await reader.read(30))
            self.ms.log(s)

            if s.startswith("b'GET / "):
                self.ms.log("Verzeichnis / erkannt. Sende...")
                await writer.awrite('HTTP/1.1 200 OK\r\n')
                await writer.awrite('Content-Type: text/html\r\n\r\n')
                await asyncio.sleep(0.1)
                await self.generateHTML(writer)
                self.ms.log("Verbindung 200 getrennt")

            else:
                await writer.awrite('HTTP/1.1 404 Not Found\r\n')
                await writer.awrite('Content-Type: text/html\r\n\r\n')
                await writer.awrite('<>')
                self.ms.log("Verbindung 404 getrennt")

            reader.close()
            await asyncio.sleep(1)
            await reader.wait_closed()
            await asyncio.sleep(0.2)
            await writer.drain()
            await asyncio.sleep(0.2)
            writer.close()
            await asyncio.sleep(0.2)
            await writer.wait_closed()

        except Exception as e:
            self.ms.log("Fehler in HTTP Loop: " + str(e))

````

The routines works fine on microsoft edge or with wget. But on Firefox the site generates successful but every time writer.close() will be called Firefox stop the site with "ERR_CONNECTION_RESET" (See Picture)

![Screenshot 2023-02-13 202737](https://user-images.githubusercontent.com/42001654/218555177-01cea7bc-dadc-4409-8921-e385ddd6e05b.png)

The asycio.sleep() methods are only for debugging and trys to get fixed.

I hope you can help me. 
Thank you alot!

**Comment by peterhinch** [comment]
You might like to look at [the uasyncio source](https://github.com/micropython/micropython/blob/41ed01f1394ea608bf9d055120d671e2765cd9b7/extmod/uasyncio/stream.py#L22). `writer.close` does nothing. It's more likely that the fault occurs on
```python
await writer.wait_closed()
```
This closes the socket. However at this point the writer has completed its work because you waited on `.drain` so I don't understand why closing the socket is problematic.

**Reply by milo1987** [reply]
Ah good to know.. then i can delete this...

But why Firefox makes a problem with that, i dont know :(
I got the hole Page, see it and when the httpstatus changes to 200 OK the page changes to connection reset.

I will try to change the http page next into a easy hello world page.

**Comment by karfas** [comment]
In my opinion it is bad practice to close sockets (or other resources) in a function that didn't open it.
Just leave the socket handling to the async server function.

You can analyze FF's behaviour using it's web console.

BTW, I don't understand all these asyncio.sleep() calls sprinkled around. I'm pretty sure asyncio will work as well without. There is no need to schedule other tasks before you `await` the completion of another function.

**Reply by milo1987** [reply]
asyncio.sleep() was just another try to get it work. On google were some solutions, so i just tried it... but that doesn't change anything... 

if i won't close the connection the firefox will never stop the connection. it's running the hole time :(
I will try to debug with the webconsole today

**Comment by milo1987** [comment, marked as answer]
OK - i found a solution.

`s = str(await reader.read(30))`

MUST for firefox be:

`s = str(await reader.read(-1))`

If you won't get everything from the clientbuffer. Firefox will stop with this error, because his connection will not be closed.

The hole code to handle the HTML Request is:

```
async def httpLoop(self, reader: asyncio.StreamReader, writer: asyncio.StreamWriter):
        try:
            self.ms.log("HTTP-Client Connected...")
            s = str(await reader.read(-1))
            self.ms.log(s)

            if s.startswith("b'GET / "):
                self.ms.log("Reading Path / . Send: ")
                await writer.awrite('HTTP/1.1 200 OK\r\n')
                await writer.awrite('Connection: close\r\n')
                await writer.awrite('Content-Type: text/html\r\n\r\n')
                await self.generateHTML(writer)
                self.ms.log("Connection 200 finished")

            else:
                await writer.awrite('HTTP/1.1 404 Not Found\r\n')
                await writer.awrite('Connection: close\r\n')
                await writer.awrite('Content-Type: text/html\r\n\r\n')
                await writer.awrite('<>')
                self.ms.log("Connection 404 finished")

            await reader.wait_closed()
            await writer.drain()
            await writer.wait_closed()

        except Exception as e:
            self.ms.log("Error in HTTP Loop: " + str(e))
```
